### PR TITLE
Add support for Key Connector OTP and account migration

### PR DESCRIPTION
**Pull request by Hinton**
## Type of change
- [ ] Bug fix
- [x] New feature development
- [ ] Tech debt (refactoring, code cleanup, dependency upgrades, etc)
- [ ] Build/deploy pipeline (DevOps)
- [ ] Other

## Objective
Adds support for the remaining Key Connector functionality. Such as using a OTP to verify actions instead of the master password. Bug fixes and support for on-boarding existing users.

## Code changes
* **bitwarden_license/src/Sso/Controllers/AccountController.cs**: Add event "LinkedSso" when creating the SSO user.
* **AccountsController.cs** Block changing email, add support for converting account to use crypto agent, refactor verifications to support either master password or OTP. And add endpoints for generating an OTP, as well as verifying it (for pure clientside actions).
* **CiphersController.cs**: Add support for OTP for the Purge action.
* **OrganizationUsersController.cs**: **TODO: Do we want this change?**
* **OrganizationsController.cs**: Add OTP fallback for action which usually requires password to verify.
* **TwoFactorController.cs**: Add OTP fallback for actions which usually requires password to verify.
* **CustomTokenRequestValidator.cs**: Add `CryptoAgentUrl` to 
* **various RequestModels**: Extend the `VerifyPasswordRequestModel` to simplify logic.
* **src/Core/Models/Api/Request/Accounts/VerifyOtpRequestModel.cs**: New request model for use with verify OTP endpoint.
* **src/Core/Models/Api/Request/Accounts/VerifyPasswordRequestModel.cs**: Added support for OTP, ensures at least one is provided.
* **src/Core/Models/Api/Response/ProfileOrganizationResponseModel.cs**: Add `UsesCryptoAgent` and `CryptoAgentUrl` to sync response, used primarily when converting existing accounts.
* **src/Core/Models/Data/SsoConfigurationData.cs**: Move logic for serializing and de-serializing `SsoConfigurationData`. Since we'll need to do operations on it outside `SsoConfig`.
* **src/Core/Models/Table/SsoConfig.cs**: Forward actions to SsoConfigurationData.
* **HandlebarsMailService.cs**: Add support for sending OTP email.
* **OrganizationService.cs**: Prevent deleting organization which uses crypto agent.
* **PolicyService.cs**: Update policy check with missing checks for MaximumVaultTimeout, and add checks to ensure single org cannot be disabled when crypto agent is enabled.
* **SsoConfigService.cs**: Prevent enabling crypto agent unless single org is enabled.
* **UserService.cs**: Add logic for converting existing account to crypto agent, support for sending OTP, validating OTP.
* **test/Api.Test/Controllers/AccountsControllerTests.cs**: Change to support actions which are checked using `VerifyPasswordOrOTPAsync`.
* **test/Core.Test/Services/PolicyServiceTests.cs**: Add tests for maximum vault timeout, and crypto agent for single org.
* **test/Core.Test/Services/SsoConfigServiceTests.cs**: Fix test failing due to invalid json.
* **src/Sql/dbo/Views/OrganizationUserOrganizationDetailsView.sql**: Add `SSOConfig` to view, used to include crypto agent details during sync.
* **util/Migrator/DbScripts/2021-11-08_00_KeyConnector.sql**: 
  * renamed from `CryptoAgent` and updated the date. This hasn't been released yet so won't affect any prod instances. The worst it will do is inconvenience devs who might have already run the old script, but it seems worth it to have accurate naming  
  * Update migration with changes to  `OrganizationUserOrganizationDetailsView`.
* **util/PostgresMigrations/** - add Postgres migrations
* **util/MySqlMigrations/** - add MySql migrations

## Before you submit
- [x] If making database changes - I have also updated Entity Framework queries and/or migrations
- [ ] I have added **unit tests** where it makes sense to do so (encouraged but not required)
- [ ] This change requires a **documentation update** (notify the documentation team)
- [ ] This change has particular **deployment requirements** (notify the DevOps team)

### CODE REVIEW

File: src/Api/Controllers/OrganizationUsersController.cs [hunk 1]
```diff
@@ -333,6 +336,16 @@ public async Task<ListResponseModel<OrganizationUserBulkResponseModel>> BulkDele
                 throw new NotFoundException();
             }
 
+            var ssoConfig = await _ssoConfigRepository.GetByOrganizationIdAsync(orgGuidId);
```
Comment (Hinton) [inline src/Api/Controllers/OrganizationUsersController.cs:339]: This is for deleting users, not org @eliykat. Not sure what we're planning to do here?

Comment (eliykat) [reply src/Api/Controllers/OrganizationUsersController.cs]: This has been removed, this is what I assumed at the time but I believe we're allowing it

File: src/Api/Controllers/AccountsController.cs [hunk 2]
```diff
@@ -490,7 +490,7 @@ public async Task Delete([FromBody]DeleteAccountRequestModel model)
 
             if (!await _userService.VerifyPasswordOrOTPAsync(user, model.Secret))
             {
-                ModelState.AddModelError("MasterPasswordHash", "Invalid password.");
+                ModelState.AddModelError(string.Empty, "Authentication failed.");
```
[the remark this answers is not in the document]
Comment (eliykat) [reply src/Api/Controllers/AccountsController.cs]: I've changed it to `User verification failed.`. It's still a bit general, but it matches the wording used in the clients, and it's less likely to be confused with auth.

File: src/Core/IdentityServer/CustomTokenRequestValidator.cs [hunk 3]
```diff
@@ -100,6 +83,37 @@ protected async override Task<(User, bool)> ValidateContextAsync(CustomTokenRequ
...
+                user.UsesKeyConnector)
+            {
+                // Apikey login
+                // KeyConnectorUrl is configured in the CLI client, just disable master password reset    
+                context.Result.CustomResponse["ResetMasterPassword"] = false;
+                return;
+            }
+
+            // SSO login
+            var organizationClaim = context.Result.ValidatedRequest.Subject?.FindFirst(c => c.Type == "organizationId");
```
Comment (eliykat) [inline src/Core/IdentityServer/CustomTokenRequestValidator.cs:105]: Do we need to add null checks in this section, e.g. what if the user is not using SSO and needs to reset their password? I assume they'll pass the guard on line 87 and then encounter errors below.

File: src/Core/IdentityServer/CustomTokenRequestValidator.cs [hunk 4]
```diff
@@ -100,6 +83,38 @@ protected async override Task<(User, bool)> ValidateContextAsync(CustomTokenRequ
...
+            // KeyConnector responses below
+
+            if (context.Result.ValidatedRequest.GrantType == "client_credentials")
+            {
+                // Apikey login
+                if (user.UsesKeyConnector) {
+                    // KeyConnectorUrl is configured in the CLI client, just disable master password reset    
+                    context.Result.CustomResponse["ResetMasterPassword"] = false;
+                }
+                return;
```
Comment (eliykat) [inline src/Core/IdentityServer/CustomTokenRequestValidator.cs:101]: I just restructured this section to make sure that apikey logins always hit this early return, even if they're not using Key Connector. Otherwise they'll continue to the SSO section below and hit nulls.

File: src/Core/Services/Implementations/SsoConfigService.cs [hunk 5]
```diff
@@ -23,6 +34,35 @@ public async Task SaveAsync(SsoConfig config)
             {
                 config.CreationDate = now;
             }
+
+            if (config.GetData().UseKeyConnector)
```
Comment (eliykat) [inline src/Core/Services/Implementations/SsoConfigService.cs:38]: `GetData()` deserializes, should we assign these calls to local variables to avoid repeatedly deserializing the same data, or is C# smart enough to cache the results?

File: src/Core/Services/Implementations/PolicyService.cs [hunk 6]
```diff
@@ -57,10 +60,24 @@ public class PolicyService : IPolicyService
                         {
                             throw new BadRequestException("Single Sign-On Authentication policy is enabled.");
                         }
+
+                        var vaultTimeout =
```
Comment (eliykat) [inline src/Core/Services/Implementations/PolicyService.cs:64]: Can we move this into its own PR? This one is big enough already.

File: test/Core.Test/Services/PolicyServiceTests.cs [hunk 7]
```diff
@@ -97,6 +98,66 @@ await sutProvider.GetDependency<IEventService>()
                 .LogPolicyEventAsync(default, default, default);
         }
 
+        [Theory, CustomAutoData(typeof(SutProviderCustomization))]
+        public async Task SaveAsync_SingleOrg_VaultTimeoutEnabled_ThrowsBadRequest([PolicyFixtures.Policy(Enums.PolicyType.SingleOrg)] Core.Models.Table.Policy policy, SutProvider<PolicyService> sutProvider)
```
Comment (eliykat) [inline test/Core.Test/Services/PolicyServiceTests.cs:102]: Move into its own PR with the new check in PolicyService, along with the other VaultTimeoutPolicy test below.

File: test/Core.Test/Services/PolicyServiceTests.cs [hunk 8]
```diff
@@ -97,6 +98,66 @@ await sutProvider.GetDependency<IEventService>()
...
+                .GetByOrganizationIdAsync(policy.OrganizationId)
+                .Returns(ssoConfig);
+
+            var badRequestException = await Assert.ThrowsAsync<BadRequestException>(
+                () => sutProvider.Sut.SaveAsync(policy,
+                    Substitute.For<IUserService>(),
+                    Substitute.For<IOrganizationService>(),
+                    Guid.NewGuid()));
+
+            Assert.Contains("KeyConnector is enabled.", badRequestException.Message, StringComparison.OrdinalIgnoreCase);
```
Comment (eliykat) [inline test/Core.Test/Services/PolicyServiceTests.cs:154]: Should we be tying our tests to the error message like this? If we change the error message, our tests will fail.

Comment (Hinton) [reply test/Core.Test/Services/PolicyServiceTests.cs:154]: It's a fairly common pattern in our code base. Ideally you would use explicit exception types which you can catch but `BadRequestException` has the advantage of wrapping to a nice json error. I think it's fine since you shouldn't change the message without intent.

File: src/Core/Services/Implementations/PolicyService.cs (hunk 6 above)
Comment (Hinton) [reply src/Core/Services/Implementations/PolicyService.cs]: Moved to https://github.com/bitwarden/server/pull/1694

File: test/Core.Test/Services/PolicyServiceTests.cs (hunk 7 above)
Comment (Hinton) [reply test/Core.Test/Services/PolicyServiceTests.cs]: Moved to https://github.com/bitwarden/server/pull/1694

File: src/Core/IdentityServer/CustomTokenRequestValidator.cs (hunk 3 above)
Comment (Hinton) [reply src/Core/IdentityServer/CustomTokenRequestValidator.cs:105]: Fair point, added null check.

File: src/Core/Services/Implementations/OrganizationService.cs [hunk 9]
```diff
@@ -908,6 +908,12 @@ public async Task UpdateLicenseAsync(Guid organizationId, OrganizationLicense li
 
         public async Task DeleteAsync(Organization organization)
         {
+            var ssoConfig = await _ssoConfigRepository.GetByOrganizationIdAsync(organization.Id);
+            if (ssoConfig?.GetData()?.UseKeyConnector == true)
+            {
+                throw new BadRequestException("You cannot delete an Organization that is using Key Connector.");
+            }
```
Comment (addisonbeck) [inline src/Core/Services/Implementations/OrganizationService.cs:915]: Lets put this in a private method to not add clutter to an already cluttered function

File: src/Api/Controllers/AccountsController.cs [hunk 10]
```diff
@@ -479,9 +511,9 @@ public async Task Delete([FromBody]DeleteAccountRequestModel model)
                 throw new UnauthorizedAccessException();
             }
 
-            if (!await _userService.CheckPasswordAsync(user, model.MasterPasswordHash))
+            if (!await _userService.VerifyPasswordOrOTPAsync(user, model.Secret))
             {
-                ModelState.AddModelError("MasterPasswordHash", "Invalid password.");
+                ModelState.AddModelError(string.Empty, "User verification failed.");
```
Comment (addisonbeck) [inline src/Api/Controllers/AccountsController.cs:516]: Why string.Empty for the key here instead of 'Secret'? There are a couple instances of this, I'm going to make a brief note on each one.

File: src/Api/Controllers/AccountsController.cs [hunk 11]
```diff
@@ -769,16 +801,13 @@ public async Task<ApiKeyResponseModel> ApiKey([FromBody]ApiKeyRequestModel model
...
+            if (!await _userService.VerifyPasswordOrOTPAsync(user, model.Secret))
             {
                 await Task.Delay(2000);
-                throw new BadRequestException("MasterPasswordHash", "Invalid password.");
-            }
-            else
-            {
-                var response = new ApiKeyResponseModel(user);
-                return response;
+                throw new BadRequestException(string.Empty, "User verification failed.");
```
Comment (addisonbeck) [inline src/Api/Controllers/AccountsController.cs:807]: Why string.Empty for the key here instead of 'Secret'?

File: src/Api/Controllers/AccountsController.cs [hunk 12]
```diff
@@ -790,17 +819,15 @@ public async Task<ApiKeyResponseModel> RotateApiKey([FromBody]ApiKeyRequestModel
...
             {
                 await Task.Delay(2000);
-                throw new BadRequestException("MasterPasswordHash", "Invalid password.");
-            }
-            else
-            {
-                await _userService.RotateApiKeyAsync(user);
-                var response = new ApiKeyResponseModel(user);
-                return response;
+                throw new BadRequestException(string.Empty, "User verification failed.");
```
Comment (addisonbeck) [inline src/Api/Controllers/AccountsController.cs:825]: Why string.Empty for the key here instead of 'Secret'?

File: src/Api/Controllers/CiphersController.cs [hunk 13]
```diff
@@ -558,9 +558,9 @@ public async Task PostPurge([FromBody]CipherPurgeRequestModel model, string orga
                 throw new UnauthorizedAccessException();
             }
 
-            if (!await _userService.CheckPasswordAsync(user, model.MasterPasswordHash))
+            if (!await _userService.VerifyPasswordOrOTPAsync(user, model.Secret))
             {
-                ModelState.AddModelError("MasterPasswordHash", "Invalid password.");
+                ModelState.AddModelError(string.Empty, "User verification failed.");
```
Comment (addisonbeck) [inline src/Api/Controllers/CiphersController.cs:563]: Why string.Empty for the key here instead of 'Secret'?

File: src/Api/Controllers/OrganizationsController.cs [hunk 14]
```diff
@@ -410,10 +410,10 @@ public async Task Delete(string id, [FromBody]OrganizationDeleteRequestModel mod
                 throw new UnauthorizedAccessException();
             }
 
-            if (!await _userService.CheckPasswordAsync(user, model.MasterPasswordHash))
+            if (!await _userService.VerifyPasswordOrOTPAsync(user, model.Secret))
             {
                 await Task.Delay(2000);
-                throw new BadRequestException("MasterPasswordHash", "Invalid password.");
+                throw new BadRequestException(string.Empty, "User verification failed.");
```
Comment (addisonbeck) [inline src/Api/Controllers/OrganizationsController.cs:416]: Why string.Empty for the key here instead of 'Secret'?

File: src/Api/Controllers/TwoFactorController.cs [hunk 15]
```diff
@@ -368,18 +368,18 @@ public async Task PostRecover([FromBody]TwoFactorRecoveryRequestModel model)
...
             {
                 throw new UnauthorizedAccessException();
             }
 
-            if (!await _userService.CheckPasswordAsync(user, masterPasswordHash))
+            if (!await _userService.VerifyPasswordOrOTPAsync(user, secret))
             {
                 await Task.Delay(2000);
-                throw new BadRequestException("MasterPasswordHash", "Invalid password.");
+                throw new BadRequestException(string.Empty, "User verification failed.");
```
Comment (addisonbeck) [inline src/Api/Controllers/TwoFactorController.cs:382]: Why string.Empty for the key here instead of 'Secret'?

File: src/Core/Models/Data/SsoConfigurationData.cs [hunk 16]
```diff
@@ -13,11 +14,25 @@ public class SsoConfigurationData
...
+
+        public static SsoConfigurationData Deserialize(string data)
+        {
+            return JsonSerializer.Deserialize<SsoConfigurationData>(data, _jsonSerializerOptions);
+        }
+
+        public string Serialize()
+        {
+            return JsonSerializer.Serialize(this, _jsonSerializerOptions);
+        }
```
Comment (addisonbeck) [inline src/Core/Models/Data/SsoConfigurationData.cs:30]: We have a generic method that accomplishes this exact thing in CoreHelpers. It may still be useful to have hardcoded type information on each specific class that has serialized data, but those methods should at least call the global generic method instead of duplicating the defaults and calls to the serializer.

File: src/Core/Services/Implementations/UserService.cs [hunk 17]
```diff
@@ -636,33 +637,54 @@ public override Task<IdentityResult> ChangePasswordAsync(User user, string maste
...
+                Logger.LogWarning("Already uses key connector.");
+                return IdentityResult.Failed(_identityErrorDescriber.UserAlreadyHasPassword());
+            }
+
+            user.RevisionDate = user.AccountRevisionDate = DateTime.UtcNow;
+            user.MasterPassword = null;
+            user.UsesKeyConnector = true;
+
+            await _userRepository.ReplaceAsync(user);
+            await _eventService.LogUserEventAsync(user.Id, EventType.User_ChangedPassword);
```
Comment (addisonbeck) [inline src/Core/Services/Implementations/UserService.cs:683]: Is this the right event?

File: src/Core/Services/Implementations/PolicyService.cs [hunk 18]
```diff
@@ -57,6 +60,12 @@ public class PolicyService : IPolicyService
                         {
                             throw new BadRequestException("Single Sign-On Authentication policy is enabled.");
                         }
+
+                        var ssoConfig = await _ssoConfigRepository.GetByOrganizationIdAsync(org.Id);
+                        if (ssoConfig?.GetData()?.UseKeyConnector == true)
+                        {
+                            throw new BadRequestException("KeyConnector is enabled.");
+                        }
```
Comment (addisonbeck) [inline src/Core/Services/Implementations/PolicyService.cs:68]: Lets put this in a private method to avoid adding more logic to this function. We could do this for all SinlgeOrg policies if desired, but not needed if that is considered out of scope.

File: src/Core/Services/Implementations/OrganizationService.cs [hunk 19]
```diff
@@ -908,6 +908,12 @@ public async Task UpdateLicenseAsync(Guid organizationId, OrganizationLicense li
...
+
```
Comment (addisonbeck) [inline src/Core/Services/Implementations/OrganizationService.cs:916]: Lets move this logic to a private method & preferabbly write a quick unit test for it. Also: why can't you delete an org that is using key connector?

File: src/Core/Services/Implementations/PolicyService.cs (hunk 18 above)
Comment (addisonbeck) [inline src/Core/Services/Implementations/PolicyService.cs:68]: Lets move this logic to a private method.

File: src/Core/Services/Implementations/SsoConfigService.cs [hunk 20]
```diff
@@ -23,6 +34,37 @@ public async Task SaveAsync(SsoConfig config)
...
+                var e = config.Enabled ? EventType.Organization_EnabledSso : EventType.Organization_DisabledSso;
+                await _eventService.LogOrganizationEventAsync(organization, e);
+            }
+
+            if (oldUseKeyConnector != useKeyConnector)
+            {
+                var e = useKeyConnector ? EventType.Organization_EnabledKeyConnector : EventType.Organization_DisabledKeyConnector;
+                await _eventService.LogOrganizationEventAsync(organization, e);
+            }
+
```
Comment (addisonbeck) [inline src/Core/Services/Implementations/SsoConfigService.cs:67]: We should write tests for these changes and break this up into smaller methods

File: src/Core/Models/Api/Request/Organizations/OrganizationSsoRequestModel.cs [hunk 21]
```diff
@@ -40,47 +39,11 @@ public class SsoConfigurationDataRequest : IValidatableObject
...
-            SpWantAssertionsSigned = configurationData.SpWantAssertionsSigned;
-            SpValidateCertificates = configurationData.SpValidateCertificates;
-            SpMinIncomingSigningAlgorithm = configurationData.SpMinIncomingSigningAlgorithm ?? SamlSigningAlgorithms.Sha256;
-            AdditionalScopes = configurationData.AdditionalScopes;
-            AdditionalUserIdClaimTypes = configurationData.AdditionalUserIdClaimTypes;
-            AdditionalEmailClaimTypes = configurationData.AdditionalEmailClaimTypes;
-            AdditionalNameClaimTypes = configurationData.AdditionalNameClaimTypes;
-            AcrValues = configurationData.AcrValues;
-            ExpectedReturnAcrValue = configurationData.ExpectedReturnAcrValue;
-        }
```
Comment (addisonbeck) [inline src/Core/Models/Api/Request/Organizations/OrganizationSsoRequestModel.cs:76]: I guess we just aren't using this anywhere? Any specific reasons it was removed on this PR?

File: src/Api/Controllers/TwoFactorController.cs [hunk 22]
```diff
@@ -351,7 +351,7 @@ public async Task<TwoFactorProviderResponseModel> PutDisable([FromBody]TwoFactor
         [HttpPost("get-recover")]
         public async Task<TwoFactorRecoverResponseModel> GetRecover([FromBody]TwoFactorRequestModel model)
         {
-            var user = await CheckAsync(model.MasterPasswordHash, false);
+            var user = await CheckAsync(model.Secret, false);
```
Comment (addisonbeck) [inline src/Api/Controllers/TwoFactorController.cs:354]: `CheckAsync` is local & widely used throughout this controller. Could we pass an object to it instead of this string value to make sure we don't have to refactor each one of these calls in the future when we inevitably need to check something else? May need to make an interface, but I think that would refactor and expand a lot nicer in the future.

**Review by addisonbeck — CHANGES_REQUESTED** [review]
Great work Thomas and Oscar 🚀

**Comment by addisonbeck** [comment]
I also noticed posted Kdf updates doesn't consider Key Connector users. Is that something we need to add logic to handle?

File: src/Api/Controllers/AccountsController.cs [hunk 23]
```diff
@@ -825,5 +852,33 @@ public async Task PutUpdateTempPasswordAsync([FromBody]UpdateTempPasswordRequest
 
             throw new BadRequestException(ModelState);
         }
+
+        [HttpPost("request-otp")]
+        public async Task PostRequestOTP()
+        {
+            var user = await _userService.GetUserByPrincipalAsync(User);
+            if (user is not { UsesKeyConnector: true })
```
[the remark this answers is not in the document]
Comment (Hinton) [reply src/Api/Controllers/AccountsController.cs:860]: It does a pattern match. We need to perform a null check, which means the alternative is `user == null || !user.UsesKeyConnector`. Rider complains that it can be replaced with a pattern match, which I personally find more readable.

File: src/Core/Models/Api/Request/Accounts/VerifyPasswordRequestModel.cs [hunk 24]
```diff
@@ -1,12 +1,27 @@
...
         public string MasterPasswordHash { get; set; }
+        public string OTP { get; set; }
+
+        public string Secret => SuppliedMasterPassword() ? MasterPasswordHash : OTP;
+
+        public IEnumerable<ValidationResult> Validate(ValidationContext validationContext)
+        {
+            if (string.IsNullOrEmpty(MasterPasswordHash) && string.IsNullOrEmpty(OTP))
+            {
+                yield return new ValidationResult("MasterPasswordHash or OTP must be supplied.");
```
[the remark this answers is not in the document]
Comment (Hinton) [reply src/Core/Models/Api/Request/Accounts/VerifyPasswordRequestModel.cs:18]: Validators in asp.net typically follows this flow. It's a trick to avoid having to compute the whole `IEnumerable` when an item often only wants to know the first. https://docs.microsoft.com/en-us/aspnet/core/mvc/models/validation?view=aspnetcore-6.0#ivalidatableobject

File: src/Core/Models/Api/Request/Accounts/VerifyOtpRequestModel.cs [hunk 25]
```diff
@@ -0,0 +1,10 @@
+﻿using System.ComponentModel.DataAnnotations;
+
+namespace Bit.Core.Models.Api
+{
+    public class VerifyOtpRequestModel
+    {
+        [Required]
+        public string Otp { get; set; }
```
[the remark this answers is not in the document]
Comment (Hinton) [reply src/Core/Models/Api/Request/Accounts/VerifyOtpRequestModel.cs:8]: I would usually agree, but this is used in the response which we do a few hacks to convert between camelCase and snake case. I'm therefore a bit hesitant to make that change.

File: src/Api/Controllers/OrganizationsController.cs (hunk 14 above)
Comment (Hinton) [reply src/Api/Controllers/OrganizationsController.cs:416]: We usually use the explicit field, since it could be either `MasterPasswordHash` or `OTP` we omit it and give a generic error.

File: src/Api/Controllers/AccountsController.cs (hunk 10 above)
Comment (Hinton) [reply src/Api/Controllers/AccountsController.cs:516]: We usually use the explicit field, since it could be either `MasterPasswordHash` or `OTP` we omit it and give a generic error.

File: src/Core/MailTemplates/Handlebars/OTPEmail.text.hbs [hunk 26]
```diff
@@ -0,0 +1,5 @@
+﻿{{#>BasicTextLayout}}
+Your email verification code is: {{Token}}
+
+Use this code to complete the protected action in Bitwarden.
```
[the remark this answers is not in the document]
Comment (danielleflinn) [reply src/Core/MailTemplates/Handlebars/OTPEmail.text.hbs:4]: I think the goal of using "protected action" was to simplify development time spent on the email. If it is easy enough to pass in a variable for the protected action (export vault, view api key, etc), let's do it. Otherwise "protected action" is acceptable.

File: src/Core/Services/Implementations/HandlebarsMailService.cs [hunk 27]
```diff
@@ -755,5 +755,20 @@ public async Task SendUpdatedTempPasswordEmailAsync(string email, string userNam
             message.Category = "UpdatedTempPassword";
             await _mailDeliveryService.SendEmailAsync(message);
         }
+
+        public async Task SendOTPEmailAsync(string email, string token)
+        {
+            var message = CreateDefaultMessage("Your Email Verification Code", email);
```
[the remark this answers is not in the document]
Comment (danielleflinn) [reply src/Core/Services/Implementations/HandlebarsMailService.cs:761]: I agree with this change. Thanks Addison

File: src/Core/Services/IUserService.cs [hunk 28]
```diff
@@ -74,5 +75,8 @@ public interface IUserService
         Task<string> GenerateSignInTokenAsync(User user, string purpose);
         Task RotateApiKeyAsync(User user);
         string GetUserName(ClaimsPrincipal principal);
+        Task SendOTPAsync(User user);
+        Task<bool> VerifyOTPAsync(User user, string token);
```
[the remark this answers is not in the document]
Comment (Hinton) [reply src/Core/Services/IUserService.cs:79]: Many of our services are growing unyielding. I'm going to mark it as out of scope for this PR, but I've had some discussions with @MGibson1 on how we can move this forward.

File: src/Core/Services/Implementations/UserService.cs [hunk 29]
```diff
@@ -21,6 +21,7 @@
 using Microsoft.AspNetCore.DataProtection;
 using Fido2NetLib;
 using Fido2NetLib.Objects;
+using Bit.Core.Models.Api;
```
[the remark this answers is not in the document]
Comment (Hinton) [reply src/Core/Services/Implementations/UserService.cs:24]: I'm hoping the .net format will help with these changes.

File: src/Core/Models/Api/Request/TwoFactorRequestModels.cs [hunk 30]
```diff
@@ -258,11 +258,7 @@ public class TwoFactorProviderRequestModel : TwoFactorRequestModel
         public TwoFactorProviderType? Type { get; set; }
     }
 
-    public class TwoFactorRequestModel
-    {
-        [Required]
-        public string MasterPasswordHash { get; set; }
-    }
+    public class TwoFactorRequestModel : VerifyPasswordRequestModel { }
```
[the remark this answers is not in the document]
Comment (Hinton) [reply src/Core/Models/Api/Request/TwoFactorRequestModels.cs:261]: Going to keep this, since we extend it in many places.

File: src/Core/Services/Implementations/PolicyService.cs (hunk 18 above)
Comment (Hinton) [reply src/Core/Services/Implementations/PolicyService.cs:68]: I think we need to look into refactoring the whole service rather than moving things to private methods. Ideally each policy should have it's own class with it's own validate method instead of having a switch case.

Comment (Hinton) [reply src/Core/Services/Implementations/PolicyService.cs:68]: Duplicate of above.

File: src/Core/Models/Api/Request/Organizations/OrganizationSsoRequestModel.cs (hunk 21 above)
Comment (Hinton) [reply src/Core/Models/Api/Request/Organizations/OrganizationSsoRequestModel.cs:76]: Not really. I touched upon the code, in this and the previous PR so it seemed a good opportunity to remove it.

File: src/Core/Services/Implementations/OrganizationService.cs (hunk 19 above)
Comment (Hinton) [reply src/Core/Services/Implementations/OrganizationService.cs:916]: It's a quick way to prevent locking out all the users using key connector.

**Comment by Hinton** [comment]
We're not currently supporting updating KDF with key connector.

File: src/Core/Services/Implementations/UserService.cs (hunk 17 above)
Comment (Hinton) [reply src/Core/Services/Implementations/UserService.cs]: Changed to `User_MigratedKeyToKeyConnector`

File: src/Core/Services/IUserService.cs (hunk 28 above)
[the remark this answers is not in the document]
Comment (addisonbeck) [reply src/Core/Services/IUserService.cs:79]: What's been the result of the discussions? I'd say at least this service break out is in scope, given that all these methods are new to this feature, but if there is a plan/ticket for tackling to the whole service as tech debt I'll concede.

File: src/Core/Models/Api/Request/TwoFactorRequestModels.cs (hunk 30 above)
[the remark this answers is not in the document]
Comment (addisonbeck) [reply src/Core/Models/Api/Request/TwoFactorRequestModels.cs]: Could the classes in question pivot to inheriting `VerifyPasswordRequestModel` instead without breaking anything?

File: src/Api/Controllers/AccountsController.cs (hunk 10 above)
Comment (addisonbeck) [reply src/Api/Controllers/AccountsController.cs:515]: Secret is an explicit field of the request model, I would say that is still valuable information for these messages.

File: src/Core/Services/Implementations/PolicyService.cs (hunk 18 above)
Comment (addisonbeck) [reply src/Core/Services/Implementations/PolicyService.cs:75]: I'm okay with leaving this as is if we are recording the tech debt somewhere that it will be prioritized, same with the UserService comments.

File: src/Core/Models/Api/Request/Accounts/VerifyOtpRequestModel.cs [hunk 31]
```diff
@@ -2,9 +2,9 @@
 
 namespace Bit.Core.Models.Api
 {
-    public class ApiKeyRequestModel
+    public class VerifyOtpRequestModel
```
Comment (addisonbeck) [inline src/Core/Models/Api/Request/Accounts/VerifyOtpRequestModel.cs:5]: > Re: Otp -> OTP 

_(we lost the original thread it seems)_

We do hacks converting between camel and snake case that effect the 3'rd+ capitals of object names? Wouldn't any hacks just apply to first capitals?

File: src/Core/Services/IUserService.cs (hunk 28 above)
[the remark this answers is not in the document]
Comment (addisonbeck) [reply src/Core/Services/IUserService.cs:79]: What's been the result of the discussions? I'd say at least this service break out is in scope, given that all these methods are new to this feature, but if there is a plan/ticket for tackling to the whole service as tech debt I'll concede.

**Review by addisonbeck — CHANGES_REQUESTED** [review]

File: src/Api/Controllers/AccountsController.cs (hunk 10 above)
Comment (Hinton) [reply src/Api/Controllers/AccountsController.cs:515]: @addisonbeck Secret is an explicit field on the backend, but this data is send to the clients, where secret doesn't exist.

File: src/Core/Models/Api/Request/Accounts/VerifyOtpRequestModel.cs (hunk 31 above)
Comment (Hinton) [reply src/Core/Models/Api/Request/Accounts/VerifyOtpRequestModel.cs]: We do those hacks on responses, so request should be safe. I've renamed the class as well (I only renamed the field before)

File: src/Core/Services/IUserService.cs (hunk 28 above)
[the remark this answers is not in the document]
Comment (Hinton) [reply src/Core/Services/IUserService.cs:79]: Going to postpone it for a separate PR, `VerifyOTPAsync` relies on `UserManager` we would end up having two services depending on the same base class which feels a bit off.

File: src/Core/Services/Implementations/PolicyService.cs (hunk 18 above)
Comment (Hinton) [reply src/Core/Services/Implementations/PolicyService.cs:75]: Added a tech debt ticket.

File: src/Core/Services/Implementations/OrganizationService.cs (hunk 19 above)
Comment (Hinton) [reply src/Core/Services/Implementations/OrganizationService.cs:912]: Added tests.

File: src/Core/Services/IUserService.cs (hunk 28 above)
[the remark this answers is not in the document]
Comment (addisonbeck) [reply src/Core/Services/IUserService.cs:79]: Thanks, Oscar, that does sound like a mess. If this is being tracked as tech debt we are all square.

**Review by addisonbeck — APPROVED** [review]
